src/data/buffer/parser/basic.lean

```diff
@@ -0,0 +1,411 @@
+/-
+Copyright (c) 2020 Yakov Pechersky. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Yakov Pechersky
+-/
+import data.list.basic
+
+/-!
+# Parsers
+
+`parser α` is the type that describes a computation that can ingest a `char_buffer`
+and output, if successful, a term of type `α`.
+This file expands on the definitions in the core library, proving that all the core library
+parsers are `valid`. There are also lemmas on the composability of parsers.
+
+## Main definitions
+
+* 'parse_result.pos` : The position of a `char_buffer` at which a `parser α` has finished.
+* `parser.valid` : The property that a parser only moves forward within a buffer,
+  in both cases of success or failure.
+
+## Implementation details
+
+Lemmas about how parsers are valid are in the `valid` namespace. That allows using projection
+notation for shorter term proofs that are parallel to the definitions of the parsers in structure.
+
+-/
+
+open parser parse_result
+
+/--
+For some `parse_result α`, give the position at which the result was provided, in either the
+`done` or the `fail` case.
+-/
+@[simp] def parse_result.pos {α} : parse_result α → ℕ
+| (done n _) := n
+| (fail n _) := n
+
+namespace parser
+
+section defn_lemmas
+
+variables {α β : Type} (msgs : thunk (list string)) (msg : thunk string)
+variables (p q : parser α) (cb : char_buffer) (n n' : ℕ) {err : dlist string}
+variables {a : α} {b : β}
+
+/--
+A `parser α` is defined to be `valid` if the result `p cb n` it gives,
+for some `cb : char_buffer` and `n : ℕ`, (whether `done` or `fail`),
+is always at a `parse_result.pos` that is at least `n`. Additionally, if the position `n` provided
+was within the size of the `cb`, then the `parse_result.pos` of the output will also be within
+the size of the `cb`.
+-/
+def valid : Prop :=
+  ∀ (cb : char_buffer) (n : ℕ), n ≤ (p cb n).pos ∧ ((p cb n).pos ≤ cb.size → n ≤ cb.size)
```

Comment:
```suggestion
∀ (cb : char_buffer) (n : ℕ), n ≤ (p cb n).pos ∧ ((p cb n).pos ≤ cb.size → n ≤ cb.size)
```
The second conjuct `(p cb n).pos ≤ cb.size → n ≤ cb.size` is exactly the opposite of what the docstring says?